### PR TITLE
Make Plants harvestable

### LINKED ISSUE #79: Make Plants harvestable

### PR DESCRIPTION
**Pull request by Sui-nyan**
- Plants can be harvested in different stages
- if fruits are harvested it goes back to pre-flower stage
- also replaced old sprite with new sprite and added one more stage 'sprouting'

closes #79

### CODE REVIEW

**Comment by PsiLotY** [comment]
I noticed a Bug: 
I let a plant grow to full and let it die. When dead i could click harvest. That changed it to the fully grown plant. Then i was able to harvest it twice once in fully grown and once in the one before to get to the vegetating state. Then i let it die in that state and could never harvest the dead plant again.

File: Plants/APlant.cs [hunk 1]
```diff
@@ -42,20 +41,63 @@ public partial class APlant : Node2D
...
+
+		AbsorbWaterFromTile();
+		ConsumeWater();
+		AbsorbSun();
+
+		if (!CheckRequirements())
+			return;
+
+		_currentDay++;
+		AdvanceStage();
```
[the remark this answers is not in the document]
Comment (Sui-nyan) [reply Plants/APlant.cs]: done

File: Plants/APlant.cs [hunk 2]
```diff
@@ -42,20 +41,63 @@ public partial class APlant : Node2D
...
+	/// Executes the growth cycle of the plant by absorbing water, consuming water, and absorbing sun.
+	/// Checks if the requirements are fulfilled to advance to the next stage.
+	/// </summary>
+	public void DoGrowthCycle()
+	{
+		if (Stage == GrowthStage.Dead)
+			return;
+
+		AbsorbWaterFromTile();
+		ConsumeWater();
```
[the remark this answers is not in the document]
Comment (Sui-nyan) [reply Plants/APlant.cs]: did it

File: Plants/APlant.cs
[the remark this answers is not in the document]
Comment (Sui-nyan) [reply Plants/APlant.cs:1]: did it

File: Plants/APlant.cs [hunk 3]
```diff
@@ -164,26 +198,6 @@ private void AbsorbSun()
 		_logger.Info(RequirementType.sun.ToString() + _currentRequirements.GetValueOrDefault(RequirementType.sun.ToString()));
```
[the remark this answers is not in the document]
Comment (Sui-nyan) [reply Plants/APlant.cs]: ok

File: Plants/Requirement.cs
[the remark this answers is not in the document]
Comment (Sui-nyan) [reply Plants/Requirement.cs:1]: ogay

File: TestScenes/APlant/APlantPrefab.tscn [hunk 4]
```diff
@@ -77,7 +89,8 @@ animations = [{
 
 [node name="APlantPrefab" type="Node2D" groups=["Plant"]]
```
[the remark this answers is not in the document]
Comment (Sui-nyan) [reply TestScenes/APlant/APlantPrefab.tscn]: ogay

File: TestScenes/APlant/TestAPlant.cs
[the remark this answers is not in the document]
Comment (Sui-nyan) [reply TestScenes/APlant/TestAPlant.cs:1]: jojo

File: Features/Dialogue/DialogueAnimation.cs [hunk 5]
```diff
@@ -1,14 +1,16 @@
...
 public partial class DialogueAnimation : Node
 {
-	private const float CharacterPerSecond = 40; // range 25 - 40 
+	private const float CharacterPerSecond = 40; // range 25 - 40
 	public bool IsPlaying { get; private set; }
 
-	[Export] private Timer _timer;
+	[Export]
+	private Timer _timer;
 	private Logger _logger;
```
Comment (PsiLotY) [inline Features/Dialogue/DialogueAnimation.cs:14]: Nitpicky, but should be after the public attributes

File: Features/Item/InteractableItem.cs [hunk 6]
```diff
@@ -1,42 +1,29 @@
...
 namespace untitledplantgame.Item;
 
-public partial class InteractableItem : Area2D, IInteractable
+public partial class InteractableItem : AInteractable
 {
 	[Export]
 	private DataContainer _dataContainer;
 
+	[Export]
+	public new string ActionName { get; private set; } = "pickup";
```
Comment (PsiLotY) [inline Features/Item/InteractableItem.cs:14]: why is this needed as export?

File: Features/VendingMachine/UI/VendingMachineUI.cs [hunk 7]
```diff
@@ -42,6 +44,15 @@ public override void _Ready()
...
+	}
+
+	private void OpenThis(VendingMachine vendingMachine)
+	{
+		GameStateMachine.Instance.SetState(GameState.Book);
+		SetVendingMachine(vendingMachine);
+		Show();
 	}
 
 	public override void _Process(double delta)
```
Comment (PsiLotY) [inline Features/VendingMachine/UI/VendingMachineUI.cs:58]: move process up according to code order

File: Scenes/TestScenes/APlant/TestAPlant.cs [hunk 8]
```diff
@@ -22,9 +22,22 @@ public override void _Ready()
 		waterSoilButton.Pressed += OnWaterSoilButtonPressed;
 		var harvestButton = GetNode<Button>("VBoxContainer/Button2");
 		harvestButton.Pressed += OnHarvestButtonPressed;
+		_label = GetNode<Label>("VBoxContainer/Label");
+		GD.Print(_label.Name);
```
Comment (PsiLotY) [inline Scenes/TestScenes/APlant/TestAPlant.cs:26]: use logger if important, else comment out or remove if just testing
(all 3 GD.Print)

File: Scenes/TestScenes/Dialogue/TestDialogue.cs [hunk 9]
```diff
@@ -1,21 +1,28 @@
...
+	private DialogueResourceObject _followupDialogue;
+
 	public override void _Ready()
 	{
 		// Prepare
-			
-			
+
+
 		GD.Print($"Current game state: {GameStateMachine.Instance.CurrentState}");
```
Comment (PsiLotY) [inline Scenes/TestScenes/Dialogue/TestDialogue.cs:21]: change to logger if neccessary

File: Scenes/TestScenes/Dialogue/TestDialogue.cs [hunk 10]
```diff
@@ -1,21 +1,28 @@
...
 		GD.Print("Trigger dialogue.");
-		
+
 		DialogueSystem.Instance.StartDialog(_exampleDialogue._dialogueId);
-		
+
 		GD.Print($"Current game state: {GameStateMachine.Instance.CurrentState}");
```
Comment (PsiLotY) [inline Scenes/TestScenes/Dialogue/TestDialogue.cs:26]: logger

File: Scenes/TestScenes/VendingMachineTest/UiOnly.cs
Comment (PsiLotY) [inline Scenes/TestScenes/VendingMachineTest/UiOnly.cs:1]: logger

**Review by PsiLotY — CHANGES_REQUESTED** [review]

**Comment by PsiLotY** [comment]
Bug is Fixed, but plants can't be harvested when they died. I feel like this would be neccessary.

File: Features/Dialogue/DialogueAnimation.cs (hunk 5 above)
Comment (Sui-nyan) [reply Features/Dialogue/DialogueAnimation.cs:14]: adjusted

File: Features/Item/InteractableItem.cs (hunk 6 above)
Comment (Sui-nyan) [reply Features/Item/InteractableItem.cs:14]: idk why maybe to see it in the inspector uring runtime? I didn't do this

File: Features/VendingMachine/UI/VendingMachineUI.cs (hunk 7 above)
Comment (Sui-nyan) [reply Features/VendingMachine/UI/VendingMachineUI.cs:58]: adjusted

File: Scenes/TestScenes/APlant/TestAPlant.cs (hunk 8 above)
Comment (Sui-nyan) [reply Scenes/TestScenes/APlant/TestAPlant.cs:26]: removed prints

File: Scenes/TestScenes/Dialogue/TestDialogue.cs (hunk 9 above)
Comment (Sui-nyan) [reply Scenes/TestScenes/Dialogue/TestDialogue.cs:21]: removed prints

File: Scenes/TestScenes/Dialogue/TestDialogue.cs (hunk 10 above)
Comment (Sui-nyan) [reply Scenes/TestScenes/Dialogue/TestDialogue.cs:26]: but also i think code specifically only for the tests are okay if not clean with print

File: Scenes/TestScenes/VendingMachineTest/UiOnly.cs
Comment (Sui-nyan) [reply Scenes/TestScenes/VendingMachineTest/UiOnly.cs:1]: siehe oben

File: Scenes/TestScenes/Dialogue/TestDialogue.cs (hunk 10 above)
Comment (PsiLotY) [reply Scenes/TestScenes/Dialogue/TestDialogue.cs]: prints for debugging the code to make it work as it should is totally fine. But when it stays, it should be commented or changed to logger

**Review by PsiLotY — CHANGES_REQUESTED** [review]
review order in DialogAnimation agian then it's good

**Comment by PsiLotY** [comment]
If the only change is going to be in the order of attributes in DialogAnimation.cs @Sui-nyan has my blessing to resolve, review and merge.

**Review by kida31 — APPROVED** [review]

**Review by PsiLotY — APPROVED** [review]
